### PR TITLE
Verify test resources by compiling them

### LINKED ISSUE #37: Verify that test resources actually compile

### PR DESCRIPTION
**Pull request by algomaster99**
Fixes #37 

I have written a bash script in this PR which compiles each Java program in the test resource directory. However, it misses two things as of now:

- [x] Rename Java files which contain a public class
- [x] If any errors are there in compiling at least 1 Java program, an exit status of non-zero should be returned.

### CODE REVIEW

**Comment by algomaster99** [comment]
@slarse In response to your comment [here](https://github.com/SpoonLabs/diffmin/issues/37#issuecomment-831097073), I did not write a python script because I did not want to maintain dependencies like [`coloroma`](https://pypi.org/project/colorama/) for it. Moreover, I thought I would have to setup a python interpreter in GHA which would take some extra time for the builds to execute.

File: scripts/verify-test-resource.sh [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+WHITE=$(tput setaf 7)
+NC=$(tput sgr0)
+RED_BG=$(tput setab 1)
+
+# Path to test resources
+TEST_RESOURCES_PATH=src/test/resources
+
+# Temporary directory for saving compiled classes
+TEMP_DIR_COMPILED_RESOURCES=temp
+mkdir -p $TEMP_DIR_COMPILED_RESOURCES
```
Comment (slarse) [inline scripts/verify-test-resource.sh:18]: Dangerous, what if someone runs this locally and there's already a directory called `temp`? Since you don't check if the directory was actually created, you could later end up deleting a preexisting directory.

Use `mktemp -d` instead, it creates a temporary directory in `/tmp` and echos the path to it.

```bash
TEMP_DIR_COMPILED_RESOURCES=$(mktemp -d)
```

File: scripts/verify-test-resource.sh [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+# Find all Java files inside the test resources directory
+find $TEST_RESOURCES_PATH -type f -name "*.java" | while read -r file
+  do
+    compile "$file"
+  done
+
+# Delete the temporary directory as it served its purpose
+echo -e "\nRemoving $TEMP_DIR_COMPILED_RESOURCES directory ..."
+rm -rf $TEMP_DIR_COMPILED_RESOURCES
+echo "Done, script execution complete."
```
Comment (slarse) [inline scripts/verify-test-resource.sh:40]: If there's a crash at any point during the script, this doesn't get executed. I'd register this in a trap, see [this SO post](https://stackoverflow.com/a/34676160)

**Review by slarse — COMMENTED** [review]
> If any errors are there in compiling at least 1 Java program, an exit status of 0 should be returned.

Then it should exit **non-zero**. 0 means everything is great, non-zero means something went wrong.

> In response to your comment here, I did not write a python script because I did not want to maintain dependencies like coloroma for it.

Is color output really that important? In any case, if it is, you can always just use ANSI escape codes for colorization. But still, I feel like that's much more than is needed here, all that's necessary is a non-zero exit status if any single Java file does not compile, along with a nice print of the path to that file.

> Moreover, I thought I would have to setup a python interpreter in GHA which would take some extra time for the builds to execute.

It takes seconds to setup a custom interpreter with the `setup-python`, and on `ubuntu-latest` there's already Python 3.6 preinstalled. Or 3.7, I don't quite remember.

File: scripts/verify-test-resource.sh (hunk 1 above)
Comment (algomaster99) [reply scripts/verify-test-resource.sh:18]: Better solution :)

**Comment by algomaster99** [comment]
> Is color output really that important?

That was the only dependency which I thought we would require. If there are more, I would want to refrain from maintaining a `Pipfile` or `requirements.txt`.

Also, in python we would have to use subprocess module to execute commands for compiling Java. We can directly do it in bash. Finally, I am less experienced in bash scripting so I thought I would take this issue as a practice to write bash script. :D

> ubuntu-latest there's already Python 3.6 preinstalled. Or 3.7, I don't quite remember

Ah, right! We would not even have to setup the interpreter in that case.

**Comment by algomaster99** [comment]
@slarse Done with one sub-task, I will fix errors in the test files and then push them here.

**Comment by slarse** [comment]
> Finally, I am less experienced in bash scripting so I thought I would take this issue as a practice to write bash script. :D

This is fine as justification. A bash script is also more portable than a Python script, but it's much harder to write safe code in bash. Much, much harder.

I recommend you set these flags in the script:

```bash
set -o errexit # if any single command exits non-zero, exit the entire script with that status
set -o nounset # don't allow use of undefined variables
set -o pipefail # if any single command in a pipe chain exits non-zero, set the exit status to that
```

**Comment by algomaster99** [comment]
@slarse 
> but it's much harder to write safe code in bash. Much, much harder.

Yes, I agree with you.

> set -o errexit # if any single command exits non-zero, exit the entire script with that status

When I put this flag, my script exits with non-zero code without even compiling a single file.

File: scripts/verify-test-resource.sh [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    echo -e "${actual_path}:${GREEN} Compiled successfully!${NC}"
+  else
+    echo -e "${RED}There were compilation errors in ${WHITE}${RED_BG}${actual_path}${NC}"
+    status=1
+    compilation_errors=$((compilation_errors+1))
+  fi
+}
+
+# Rename file according to the first public class inside
+get_and_rename_file() {
```
Comment (algomaster99) [inline scripts/verify-test-resource.sh:53]: This function checks for the **first** public class inside the java file and copies the file to `/tmp` directory with the name of the public class. Let me know if this is the best way or not.

File: scripts/verify-test-resource.sh [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+  local file=$1
+  REGEX_PATTERN="(?<=public\sclass\s)[A-Z$_][\w$]*"
```
Comment (algomaster99) [inline scripts/verify-test-resource.sh:55]: This is the regex pattern I have written to search for the name of the public class.

File: scripts/verify-test-resource.sh [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+# Colours for stdout
+RED=$(tput setaf 1)
+GREEN=$(tput setaf 2)
+WHITE=$(tput setaf 7)
+NC=$(tput sgr0)
+RED_BG=$(tput setab 1)
+
+# Path to test resources
+TEST_RESOURCES_PATH=src/test/resources
```
Comment (slarse) [inline scripts/verify-test-resource.sh:17]: This is dependent on the script being executed from the root of the project. Can be solved like so:

```suggestion
TEST_RESOURCES_PATH="$(git rev-parse --show-toplevel)/src/test/resources"
```

File: scripts/verify-test-resource.sh [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+# Temporary directory for saving compiled classes
+TEMP_DIR_COMPILED_RESOURCES=$(mktemp -d)
+echo "Created $TEMP_DIR_COMPILED_RESOURCES for saving compiled files temporarily."
+echo -e "It will be removed after the script has executed with any exit code.\n"
```
Comment (slarse) [inline scripts/verify-test-resource.sh:22]: As you don't have `errexit` enabled, you must check if the temporary directory was actually created. If it wasn't, you might have just an empty string in `TEMP_DIR_COMPILE_RESOURCES`, and then doing `rm -rf "$TEMP_DIR_COMPILE_RESOURCES"` wouldn't be great...

File: scripts/verify-test-resource.sh [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+      local new_file_path="${TEMP_DIR_COMPILED_RESOURCES}/${new_file_name}"
+      compile "$new_file_path" "$file"
+  fi
+}
+
+# Record changes to global variables
+shopt -s lastpipe
+
+# Find all Java files inside the test resources directory
+find $TEST_RESOURCES_PATH -type f -name "*.java" | while read -r file; do
```
Comment (slarse) [inline scripts/verify-test-resource.sh:77]: I'd use a for loop here. And don't forget to quote variables, spaces in file paths might otherwise cause havoc!

```suggestion
for file in $(find "$TEST_RESOURCES_PATH" -type f -name "*.java"); do
```

File: scripts/verify-test-resource.sh [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+  get_and_rename_file "$file"
+  file_count=$((file_count+1))
+done
+
+if [[ $status -eq 0 ]]
+  then
+    echo -e "\n${GREEN}All [${file_count}/${file_count}] files were compiled successfully!${NC}"
```
Comment (slarse) [inline scripts/verify-test-resource.sh:84]: `printf` is more portable than `echo -e`. Doesn't really matter here as this script is specifically to run on GHA Ubuntu, but I'd prefer `printf` over `echo -e`.

For the long explanation, see [this SO post](https://unix.stackexchange.com/a/65819)

File: scripts/verify-test-resource.sh (hunk 4 above)
Comment (slarse) [reply scripts/verify-test-resource.sh:55]: Actually, I'm surprised the regex works for the package private classes in the CI run, because it shouldn't. And it doesn't on my machine.

**Review by slarse — CHANGES_REQUESTED** [review]
This is a nice start and a very tidy script, but it's still quite shaky. I also feel the need to mention that this is _much more_ than what's actually needed here. I'm guessing you had some fun building this, and it's fine building stuff too complexly sometimes just for the fun of it, but you might want to consider dumbing this down a bit.

For example, you don't need to output which files _do_ compile compile. We only care about which files don't compile, and even then, you don't need to output all of them. It would be fine to just output the _first_ file that doesn't compile, and then exit non-zero. If all files compile, then there's no need to output anything at all. We don't expect this to fail often, and when it does it shouldn't fail for many files.

Of course, you're free to make this choice yourself. If you want the nice output, then keep it, but it might cause headaches down the road!

On a side note, the colorized output of which file did not compile (red background) is almost impossible to read for me, and that's the only information I'm actually interested in :)

File: scripts/verify-test-resource.sh (hunk 7 above)
Comment (algomaster99) [reply scripts/verify-test-resource.sh:77]: Is there any reason for using for loop apart from clarity? Is there some problem with piping?

File: scripts/verify-test-resource.sh (hunk 6 above)
Comment (algomaster99) [reply scripts/verify-test-resource.sh:22]: Okay, this makes sense. If the directory is not created, I will exit the script with status code `1`.

File: scripts/verify-test-resource.sh (hunk 4 above)
Comment (algomaster99) [reply scripts/verify-test-resource.sh:55]: I think the Java file name should match the **public class** name only. If there is not public class/interface inside the file,  the name can be any. The regex here does nothing if it doesn't find a public class. I made it that way because `javac` was compiling a file with one package private class even though the name did not match.

**Comment by algomaster99** [comment]
>  you don't need to output which files do compile compile.

Completely agreed.

>  It would be fine to just output the first file that doesn't compile, and then exit non-zero. If all files compile, then there's no need to output anything at all. We don't expect this to fail often, and when it does it shouldn't fail for many files.

I feel otherwise about this. Whenever we write a test case, we would write at least two files - `prev.java` and `new.java`. There can be compilation errors in both so I wouldn't want to find the errors in both files by running the script twice (first run will throw errors for `prev.java` and after fixing them, second run will throw errors for `new.java`). I think it would be better to give the errors all at once and since we do not expect _a lot_ of files to fail compilation, the output won't be hard to follow.

> If you want the nice output, then keep it, but it might cause headaches down the road!

The purpose of the script is just to check if all test resources compile and I get why we don't need to inundate it with `printf` statements just to log "what is happening in the script". In spite of that, I think we should at least add a summary statement at the end like "2/8 files were not compiled", "All files successfully compiled", etc. This would make it easier for other contributors to follow and they might be able to fix the errors by seeing the GHA job output themselves.

File: scripts/verify-test-resource.sh (hunk 4 above)
Comment (slarse) [reply scripts/verify-test-resource.sh:47]: Ah, yes, you're right, the file naming convention only applies to a public type. Funny, I didn't know that!

That leaves 5 ways in which the regex breaks :)

With that said, the very easiest solution would be to just... not have public types in the test files? :)

File: scripts/verify-test-resource.sh (hunk 7 above)
Comment (slarse) [reply scripts/verify-test-resource.sh]: Just for clarity.

That post is right, though, the for loop will wail if the re are spaces in the file paths. Keep your version!

File: scripts/verify-test-resource.sh (hunk 4 above)
Comment (algomaster99) [reply scripts/verify-test-resource.sh:47]: > That leaves 5 ways in which the regex breaks :)

4, actually. `\w` contains the digits from `0-9` as well but the i18n is not taken care of so yeah, it might break there. Anyway, modifying this RegEx did not seem feasible as the lookbehinds need to be of fixed width but we have variable length keywords like - `class`, `interface`, etc. So I just decided to go with solution no. 3.

> just... not have public types in the test files? :)

I thought of that too but that is more of a workaround than a solution. Suppose we would want to test for an Update Operation which updates the access modifier from `private` to `public`. Our test resources won't compile in that case. The 3rd solution which you proposed works quite well and seems most elegant to me :)

**Comment by slarse** [comment]
```
basename "$file" | cut -d '_' -f 2
```

File: scripts/verify-test-resource.sh [hunk 9]
```diff
@@ -0,0 +1,71 @@
+#!/usr/bin/env bash
+
+set -o nounset # don't allow use of undefined variables
+set -o pipefail # if any single command in a pipe chain exits non-zero, set the exit status to that
+set -o errexit # if any single command exits non-zero, exit the entire script with that status
```
Comment (algomaster99) [inline scripts/verify-test-resource.sh:5]: @slarse Had one confusion here. My script **does not** exit when at least one resource fails to compile. Although this is desired, the script should exit with a non-zero code, right?

Comment (slarse) [reply scripts/verify-test-resource.sh:5]: I'm not sure what you mean by the script not exiting. Do you mean that it doesn't immediately exit when the first resource fails to compile? Then that's because you've got it in an `if [[ ]]`, the exit status is "consumed" the conditional.

Which is what you wanted? So you can report all failed files at the end of the script?

Comment (algomaster99) [reply scripts/verify-test-resource.sh:5]: > Do you mean that it doesn't immediately exit when the first resource fails to compile?

I meant exactly this. :)

> Then that's because you've got it in an if [[ ]], the exit status is "consumed" the conditional.

Got it.

> Which is what you wanted?

Yes, that's exactly what I wanted but I thought I would have to write some extra code for it but it just worked out for me.

Comment (slarse) [reply scripts/verify-test-resource.sh:5]: Okay, then all is good :)

File: src/test/java/com/diffmin/AppTest.java [hunk 10]
```diff
@@ -35,6 +35,10 @@
 
     public static final Path DELETE_UPDATE_PATCHES = RESOURCES_BASE_DIR.resolve("delete+update");
 
+    private static final String prevPrefix = "PREV";
+
+    private static final String newPrefix = "NEW";
```
Comment (slarse) [inline src/test/java/com/diffmin/AppTest.java:40]: These are constants, and thus have a special naming convention of `UPPER_CASE`. E.g. `PREV_PREFIX`.

File: src/test/java/com/diffmin/AppTest.java [hunk 11]
```diff
@@ -73,7 +77,17 @@
...
+            List<Path> pathList = paths.collect(Collectors.toList());
+            Map<String, Path> testPathMap = new HashMap<>();
+            for (Path testPath : pathList) {
+                if (testPath.getFileName().toString().startsWith(AppTest.prevPrefix)) {
+                    testPathMap.put(AppTest.prevPrefix, testPath);
+                } else if (testPath.getFileName().toString().startsWith(AppTest.newPrefix)) {
+                    testPathMap.put(AppTest.newPrefix, testPath);
+                }
+            }
+            return new TestResources(testPathMap.get("PREV"), testPathMap.get("NEW"), parent);
```
Comment (slarse) [inline src/test/java/com/diffmin/AppTest.java:90]: Don't forget to use your constants.

**Review by slarse — CHANGES_REQUESTED** [review]
Just a few minor notes, then we merge.

On a side note, the loop in which you collect the paths to the prev and new files can be simplified to not use neither loop nor map. It _works fine_, though, so let's leave and talk about other potential approaches there during the next meeting :)

**Comment by algomaster99** [comment]
> On a side note, the loop in which you collect the paths to the prev and new files can be simplified to not use neither loop nor map. It works fine, though, so let's leave and talk about other potential approaches there during the next meeting :)

I was skeptical about the design, to be honest, as I thought there must be a shorter way to do the same thing.

Do you have any suggestions where we would should keep track of such notes? I usually use Google Keep but that's just for personal use.

File: src/test/java/com/diffmin/AppTest.java (hunk 11 above)
Comment (algomaster99) [reply src/test/java/com/diffmin/AppTest.java:90]: Oops. I will fix this.

**Comment by slarse** [comment]
> > On a side note, the loop in which you collect the paths to the prev and new files can be simplified to not use neither loop nor map. It works fine, though, so let's leave and talk about other potential approaches there during the next meeting :)
> 
> I was skeptical about the design, to be honest, as I thought there must be a shorter way to do the same thing.
> 
> Do you have any suggestions where we would should keep track of such notes? I usually use Google Keep but that's just for personal use.

I keep notes locally. If you want to synchronize notes, we could use a permanent issue on the issue tracker (can be e.g. this closed one https://github.com/SpoonLabs/diffmin/issues/19, so other people don't have to stumble onto our ramblings).

**Review by slarse — APPROVED** [review]
So, we worked a lot more on this one that I would have anticipated, but I think we both learned a thing or two so it's all good!
